<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:macro name="diffbot_ros2_control" params="name use_sim_time:=false use_mock_hardware:=false prefix:=''">

    <ros2_control name="${name}" type="system">

      <xacro:unless value="${use_sim_time}">
        
        <xacro:unless value="${use_mock_hardware}">
          <hardware>
            <plugin>zebra/ZebraSystemHardware</plugin>
            <param name="loop_rate">30</param>
            <param name="device">/dev/ttyACM0</param>
            <param name="baud_rate">115200</param>
            <param name="timeout_ms">1000</param>
            <param name="enc_counts_per_rev">2800</param>
          </hardware>
        </xacro:unless>

        <xacro:if value="${use_mock_hardware}">
          <hardware>
            <plugin>mock_components/GenericSystem</plugin>
            <param name="calculate_dynamics">true</param>
          </hardware>
        </xacro:if>
      </xacro:unless>

      <xacro:if value="${use_sim_time}">
         <hardware>
          <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>
      </xacro:if>

      <joint name="${prefix}left_wheel_back_joint">
        <command_interface name="velocity"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="${prefix}right_wheel_back_joint">
        <command_interface name="velocity"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="${prefix}right_wheel_front_joint">
        <command_interface name="velocity"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="${prefix}left_wheel_front_joint">
        <command_interface name="velocity"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

    </ros2_control>

  </xacro:macro>

</robot>